### ISSUE TITLE
import / export of user data

### ISSUE THREAD

**Issue by leuc** [pixelfed/pixelfed#814]
# A user wants to import/export data between (pixelfed) instances

Included data:
- User profile
- Uploaded images
- Image descriptions
- Image likes
- Comments on images
- Followers / Following

### What intermediate format should be used?

- [ ] JSON for direct migration between Pixelfed instances?
- [ ] What JSON namespace?
- [ ] Also export a existing XML namespace for a more generic export?
- [ ] Follower / Following data compatible with other Fediverse Software?

### Federated data

- [ ] Exported comments and like data must include full reference to source instance user id (@user@example.org) of the person who wrote comment or liked the post. 
- [ ] However, the exporting user might want to change identity between instances. Exported data must clearly define what the user id was, so it can be mapped to a new id on import.
- [ ] Import should not trigger federation?
- [ ] Like/comment data could easily be faked? Verification on import? How?
- [ ] Direct transfer between instances without the need of manual export / import? (Requires that both instances are available, which might **not** be the case in such situations.)

### Images

User does not want to import/export images individually

- [ ] Images could be packed into multi-volume archives?
- [ ] Destination instance needs to know about the existence of all images before the upload?
- [ ] Maintain or renew all image URL hashes on import?

**Comment by trwnh** [comment]
# 1) Format

## 1.1-1.3) JSON for direct migration between Pixelfed instances? namespace? XML for generic export?
Export and import should be in ActivityPub JSON format, to allow possible import into formats other than Pixelfed's internal database representation.
- The namespace would be W3C's activitystreams
- XML is not required at all for generic import/export -- ActivityPub is a standardized format already.

## 1.4) Follower/following compatibility?
Follower/following is not compatible with any other software right now because identity of Actors are dereferenced internally and ID system is up to each individual project to implement.

It is theoretically possible to reconstruct data perfectly with an offline import if you have
a) the private key (to establish ownership of content/account)
b) some globally unique identifier that is signed by the private key (to establish it is the same content/account),
c) the following/follower lists, so that you can send an Update/Move of your Actor (and content?) to them with your new location (but this requires both softwares to support Update/Move of necessary fields, like `id` and `url`).

It is also possible to do an online import with just a) and b), if the original server is still online -- the original server can provide c) for you. It's also possible to drop b), since no agreed-upon global UID standard exists within ActivityPub; the expectation is that each implementation will choose its own internal system for identifying/deduplicating accounts. But this makes it harder to identify/deduplicate accounts consistently across softwares or servers.

The last 3 paragraphs are explained a bit more in-depth at #216 but this is what you should know, at least.

At minimum, you can export following list only, and have a new account re-follow all old accounts that you were following.

# 2) Federated data

## 2.1) Exported comments/likes
This can also be in ActivityPub JSON documents, but the expectation is that those activities are stored by the server of the Actor, not necessarily by your own server. You can still export a list of account names at the time of export -- there is no way to perfectly maintain the identities of each commenter or liker, and no way to import at all; the Actor's server must Update its own records or Move your posts to point to your posts' new locations.

## 2.2) Changing identity between instances
See #216 

## 2.3) Import should not trigger federation
This is open to interpretation; the data and activities can be imported but with prior dates. Each software must negotiate data transfer of older content -- either your software redelivers new copies of the content, or you send Update activities of the new locations, or the remote software backfills statuses based on its own fetching strategies -- see https://github.com/tootsuite/mastodon/issues/34 for more.

## 2.4) Verification on import
See 2.1 -- this would be done by checking signatures on your own content (to ensure the private key matches the content), and then sending signed Update/Move to every other Actor involved.

## 2.5) Direct transport between instances
See #216 or the "online import" strategy described in 1.4

# 3) Images

## 3.1) Multi-volume archives
I don't see a particular reason why archives should specify a type of volume. This does however add the implication that software must support both partial import (append) and full import (overwrite).

## 3.2) Establish existence of images
Same as 2.4

## 3.3) Maintain image URLs
Same as #216 or 2.3/2.4.

**Comment by bitinn** [comment]
Adding my vote to this: this is kinda a big deal, I just realize we can even import from instagram backup but doesn't offer a way to backup our own.

Having this is a pretty big step towards "allow users to own their data", not to mention easy data migration between instances.

**Comment by Goldmaster** [comment]
What about the idea to be able to sync from Instagram to pixelfed? That way I can still Hootsuite. The idea is to get Hootsuite to support pixelfed for those wanting Instagram alternatives.

**Comment by xplosionmind** [comment]
Hello, sorry for dropping in our of the blue, but…

is this feature planned to be developed anytime soon? I am really looking forward to it.

**Comment by cloo** [comment]
Extremely important request for user freedom, so just doubling down.

**Comment by KaKi87** [comment]
Hello @dansup,
Any news on this ?
Thanks

**Comment by h-2** [comment]
Wait, do you mean that I can import from Instagram but not from another Pixelfed instance? o_O

**Comment by cloo** [comment]
I think devs have a different idea about priorities. I don't even notice them reacting much to github issues.

---

![image](https://github.com/pixelfed/pixelfed/assets/651760/16e9e786-2cdc-4986-ad17-61f08ddc2cd7)

---

![image](https://github.com/pixelfed/pixelfed/assets/651760/a087f691-dbeb-4696-9d39-3a99eaa09636)

I see two signs of a short-term project: one man band and feature-creep (bells'n'whistles often prioritized against strategic strength). I hope @dansup will find ways to attract more collaborators before it becomes unmanageable.

**Comment by dansup** [comment]
> I think devs have a different idea about priorities. I don't even notice them reacting much to github issues.
> 
> ![image](https://user-images.githubusercontent.com/651760/248739753-16e9e786-2cdc-4986-ad17-61f08ddc2cd7.png)
> 
> ![image](https://user-images.githubusercontent.com/651760/248738696-a087f691-dbeb-4696-9d39-3a99eaa09636.png)
> 
> I see two signs of a short-term project: one man band and feature-creep (bells'n'whistles often prioritized against strategic strength). I hope @dansup will find ways to attract more collaborators before it becomes unmanageable.

I'm working on IG Import right now, it may not seem like I respond to issues immediately, and usually it's because the reporter mentioned in in our Matrix or Discord and I ship a fix without closing or responding to relevant issues.

Will work on that, and I'm trying to make the project more encouraging to contributors.

**Comment by cloo** [comment]
Thanks @dansup ! Don't get me wrong, you are a hero, and I'm really rooting for prosperity of Pixelfed. This is why I'm wasting my time reporting issues and also moods on the user side. Be careful not to burnout too, this is a security vulnerability for all the community!

**Comment by andreagrandi** [comment]
My two cents: I think a feature to export all the data (including photos, comments, like etc...) should have an higher priority, because it could help all the Pixelfed instances to be GDPR compliant https://gdpr-info.eu/art-20-gdpr/

In particular:
>The data subject shall have the right to receive the personal data concerning him or her, which he or she has provided to a controller, in a structured, commonly used and machine-readable format

Also: Pixelfed instances are run by volunteers and they may disappear overnight. Being able to do a backup of all the data is quite important.

**Comment by pylapp** [comment]
Any news about exporting data for users with more than 500 statuses? Any tips while waiting for a new feature? 🫶

**Comment by gidTT** [comment]
I'd like the import/export feature for the reasons mentioned above. additionally it might be usueful for bulk uploading.

**Comment by MatKlein** [comment]
I'm currently facing the problem of having to switch instances. I have backed up all the data from my old instance and can only find the option to import data from Instagram on the new instance. Is there actually no way to move existing data between two Pixelfed instances?

**Comment by bigjdunham** [comment]
Bump, because I'd really like to see this happen. Needing to migrate from one instance to another because my old instance is stuck at old version.

**Comment by dansup** [comment]
We do support IG Import currently, but we need to improve our data export and fix the account migration bug. 

I will keep this issue open until it is fixed.

**Comment by grsharp** [comment]
Is there any update on this feature? I'm looking to migrate to another instance as my current server is becoming increasingly unstable. It would be great to take my archive of photos with me.

**Comment by ghost** [comment]
Yeah, I've more or less stopped using Pixelfed because not having a functional data export is a bit "shades of Meta / Instagram."

**Comment by pylapp** [comment]
Any news about this feature @dansup @shleeable?

I would like to have a backup of my statuses and collections, but it seems the feature is possible for less than 500 publications and I have much more.
If this feature is not ready or planned yet, could you share if possible some documentation to help me make kind of script making such backups?

Thank you a lot :)

<img width="300" alt="screenshot" src="https://github.com/pixelfed/pixelfed/assets/7559007/42df613e-f9e2-448a-8a15-351671affd36">

**Comment by davidoclubb** [comment]
Just adding my voice here. My current instance is being closed in Feb 2025 and I will have to manually export and re-upload all my images to a new server. Which will also (presumably) mess up the timestamps etc 🙏

I'm willing to pay a bounty to contribute to costs for this - £100 in case anybody else can add to it.

**Comment by nesges** [comment]
I'm honestly very surprised that neither full migration nor archive download are implemented in Pixelfed

**Comment by sgrigson** [comment]
Your best bet is to use [Posty](https://posty.1sland.social/pixelfed) to upload your Pixelfed archive (account.json and outbox.json) because ***there are NO images contained in your export***. If your old Pixelfed server has gone down, all that data is now gone.

Using Posty, you'll at least be able to download those images from your Pixelfed instance, and convert it into a Mastodon archive which does contain your data, and optionally to then use Posty to convert it to a static HTML site after that.

Unfortunately, you're going to lose your entire reply structure. `inReplyTo` is not populated, as such all your posts in the export, including your replies, look like top-level posts rather than maintaining the reply structure.

But it's better than nothing, for now.

**Comment by sgrigson** [comment]
Note that I've filed an issue about the missing reply structure here: https://github.com/pixelfed/pixelfed/issues/5409

**Comment by sgrigson** [comment]
Pixelfed has been updated to create a `pixelfed-statuses.json` instead of `outbox.json`. The biggest advantage is preserving the reply structure. I've updated [Posty - Pixelfed](https://posty.1sland.social/pixelfed) to accept either outbox format.

**Comment by davidoclubb** [comment]
@sgrigson Thankyou very much for your comments here, and your practical support! It has arrived just in time for me to avoid the manual hard labour of exporting each image individually! 🙏🤝

**Comment by gidTT** [comment]
I guess I'm missimg smth, but i do not see any possibility to export my images from pixelfed.de

**Comment by sgrigson** [comment]
> I guess I'm missimg smth, but i do not see any possibility to export my images from pixelfed.de

That's why I mentioned [Posty](https://posty.1sland.social) above. Posty will convert your Pixelfed archive into a Mastodon archive, and that includes downloading the missing images.

**Comment by gidTT** [comment]
thanks, i got account.json but pixelfed.de's "settings/data export/statuses" never completes download. then
504 Gateway Time-out

also i don't see how to get outbox.json

**Comment by sgrigson** [comment]
> thanks, i got account.json but pixelfed.de's "settings/data export/statuses" never completes download. then 504 Gateway Time-out
> 
> also i don't see how to get outbox.json

So `outbox.json` is what the 'statuses' exported before a recent change to pixelfed, and now instead of an outbox.json it returns pixelfed-statuses.json. Posty supports both, but yeah if you can't get a download of either of those, that is going to pose a problem.

**Comment by edmael** [comment]
Hello, is there any news about this?
We are closing our instance and we've suggested users to export their data using @sgrigson tool, but some of them are getting 504'd.

Is there any date about an official implementation of the migration feature?

**Comment by sgrigson** [comment]
> Hello, is there any news about this? We are closing our instance and we've suggested users to export their data using [@sgrigson](https://github.com/sgrigson) tool, but some of them are getting 504'd.

Yeah, it's possible for very large archives in which we have to download hundreds or thousands of images, the server might get overloaded.

If only the images were included in the export.

**Comment by Kaio20** [comment]
I think as many others have pointed out having no export feature for photos is bad. I would add to the bounty of @davidoclubb my 50€ as I think not being able to download my images automatically seems like a big risk in starting/ maintaining a long running archive like my one with close to 700 images.

**Comment by jrossstocholm** [comment]
> I think as many others have pointed out having no export feature for photos is bad. I would add to the bounty of [@davidoclubb](https://github.com/davidoclubb) my 50€ as I think not being able to download my images automatically seems like a big risk in starting/ maintaining a long running archive like my one with close to 700 images.

I am not sure how this bounty thing would work - but I would also like to contribute 50€. If the argument for using #fediverse services is "control of your data", I think full export including images/videos is a crucial piece missing.

Also - I saw that there is a "Monitization" feature planned for Q4 2026. If we want to attract content creators they will need a way to back up their content. It is already built-in on Instagram. That this is missing in pixelfed is kindof weird 😁❤️

**Comment by grossermensch** [comment]
This might also be an issue with DSGVO compliance in Europe for server admins. They should be able to export (as well as delete) all user data on request of the user.

**Comment by SievaFliesAlone** [comment]
I have been on pixelfed.social for a few years. At the moment, a few friends and I want to launch our own instance, however not having an export-import functionality is a real barrier for me.

Being able to export your data is a functionality that services have to comply with by law in many jurisdictions.

**Comment by grsharp** [comment]
Any update on this issue? Looks like it's been open for 7 years! There was a post from @dansup showing some progress but I don't think it was ever released.

**Comment by KaKi87** [comment]
And here I thought as I read the email that it was finally *completed*.

**Comment by raucao** [comment]
Maybe someone could state the status explicitly? GitHub says "closed this as completed", which suggests that you can now export and import the user data. Is that correct?

**Comment by KaKi87** [comment]
Well it's clear once you click *View on GitHub* :

> `shleeable` mentioned this
> `[CLEANUP 2026] Consolidated ActivityPub/Federation issues #6713`

It was just disappointed after seeing the email saying *completed*.

**Comment by SievaFliesAlone** [comment]
I'm not quite sure how consolidating tens of different issues into one mega-issue can help.

To me this this looks like #814 was put in a cardboard box labelled “FEDI STUFF” which was tossed into a basement to be later discovered by future generations.

**Comment by andypiper** [comment]
I don't think it is unreasonable to use a tracking issue for related topics - it can make sense to do that consolidation as a jumping-off point for related tasks. 

However, I also don't think it helps for the individual linked issues to be _closed as completed_ if they are _not completed_. You could close if not planned (#4307 #4279 [^1]): doing that is a helpful statement of project intent. Also fine close as duplicate if the issue exists in a to-be-completed state elsewhere. Closing _as completed_ when this is an important feature that is not completed, sends a very strange signal.

[^1]: this means I can't use Pixelfed embeds, and have removed them from my sites, unless someone comes up with a third-party solution, but 🤷🏻

**Comment by trwnh** [comment]
you could also not close the issues while still having them organized into a tracking issue or project or whatever.
